stan/math/rev/functor/coupled_ode_system.hpp

```diff
@@ -438,53 +112,62 @@ struct coupled_ode_system<F, var, var> {
    *    expected number of derivatives, N.
    */
   void operator()(const std::vector<double>& z, std::vector<double>& dz_dt,
-                  double t) const {
+                  double t) {
     using std::vector;
 
+    dz_dt.resize(size());
+
     // Run nested autodiff in this scope
     nested_rev_autodiff nested;
 
-    vector<var> y_vars(z.begin(), z.begin() + N_);
+    Eigen::Matrix<var, Eigen::Dynamic, 1> y_vars(N_);
+    for (size_t n = 0; n < N_; ++n)
+      y_vars.coeffRef(n) = z[n];
 
-    vector<var> dy_dt_vars = f_(t, y_vars, theta_nochain_, x_, x_int_, msgs_);
+    Eigen::Matrix<var, Eigen::Dynamic, 1> f_y_t_vars
+        = apply([&](auto&&... args) { return f_(t, y_vars, msgs_, args...); },
+                local_args_tuple_);
 
-    check_size_match("coupled_ode_system", "dz_dt", dy_dt_vars.size(), "states",
+    check_size_match("coupled_ode_system", "dy_dt", f_y_t_vars.size(), "states",
                      N_);
 
-    for (size_t i = 0; i < N_; i++) {
-      dz_dt[i] = dy_dt_vars[i].val();
-      dy_dt_vars[i].grad();
+    for (size_t i = 0; i < N_; ++i) {
+      dz_dt[i] = f_y_t_vars.coeffRef(i).val();
+      f_y_t_vars.coeffRef(i).grad();
```

Comment:
Is this why the nullptr check needs to happen in grad? If one of those vars can be nullptr that sounds like a bug more than anything.

Also am I reading this right that this need to take the adjoints wrt to all previous vars? tbh I'm pretty confused by this whole loop

Reply:
> Is this why the nullptr check needs to happen in grad?

Nah, but I removed that from this pull.

> Also am I reading this right that this need to take the adjoints wrt to all previous vars?

The forward ODE sensitivity equations for parameters are something like:

```
s = dy/dtheta
ds/dt = df/dy * s + df / dtheta
```

The sensitivity equations for initial conditions are something like:
```
a = dy/dy0
da/dt = df/dy * a
```

All that is happening in here, so it is rather confusing.

It's explained in the class docs.

Reply:
Oh yeah and the coupled_ode_system is like the combined ODE system `c(y, s, a)` so it's all getting put together there.

Reply:
:eye: